Express LTS strategy (#199)
expressjs/discussions: docs/LTS-strategy.md

# Express LTS Strategy

## Goals

* Maintain strict Node.js version support within a single Express major release.
* Provide good ergonomics and developer experience for Express maintainers
* Provide good experience for users who want to keep their Express projects that up-to-date

> * Avoid stagnation and blocking innovation that is good for the end users

Reviewer: I think folks might read this as "in order of priority". I know that is not intended, but I wonder if we should move "Provide good experience for users who want to keep their Express projects that up-to-date" to the top?

Author: addressed

## Non-Goals

* Only support the active LTS releases of Node.js

> * Support the widest possible set of Node.js versions. We aim to maintain a healthy balance between backwards compatibility and toolchain modernization.

Reviewer: Love this section, especially this line! 🎉

Reviewer 2: I'm not actually sure this makes sense - a healthy balance imo is the same thing as "widest possible set". If a node version prevents using modern toolchains, for example, then that's not a version that'd be included in that set.

Author: @ljharb I'll think about possible wording improvements here, but the intent is to emphasize that we only want to support a set of versions that do not contradict the rest of the goals defined. This is very important to make explicit in the context of Express specifically, as it has a long-standing tradition of preserving the compatibility to the detriment of everything else.

Reviewer 3: +1 this is a great way to explain it

Reviewer 2: I don’t think any of the detriments you refer to have anything to do with refusing to drop compat for node, but with other things - but i look forward to an alternative phrasing.

Reviewer: > a healthy balance imo is the same thing as "widest possible set"

As other said, I disagree with the wording but maybe your meaning hinges on the definition we use for "possible"?

Like if we said "we want to ship diagnostic channels so it is not possible to support versions before them" (I am trying to make a hypnotically based on a reality from the past https://github.com/pillarjs/router/pull/96) does that fit your definition? We would drop the old node versions which dont support it in the next major so we could ship support without a lot of compatibility code.

Reviewer 2: Yes, of course it fits my definition. If we need a feature that requires dropping a version, then that version is no longer possible to support.

I also especially like the implied path of "ship support with compat code, and then drop the compat code and the node versions that need it in the next major" - that's a *very* user-friendly way to do a major bump.

> * Support an extended set of Node.js versions to the detriment of all other factors. We aim to maintain a healthy balance between backwards compatibility and toolchain modernization

Author: @ljharb @wesleytodd @UlisesGascon Please review the updated wording

Reviewer 2: How about this:
```suggestion
* Prioritizing "support older Node.js versions" over other factors. We aim to maintain a healthy balance between backwards compatibility and toolchain modernization
```

Author: sure, that works!

### Rationale behind dropping support for old Node.js versions

* Avoid blocking Node.js core from making progress and removing deprecated parts of the codebase
* Avoid getting outdated bug reports or security issues

> * Keep options for adopting newer patterns and tooling open

Reviewer: ```suggestion
* Keep options open for adopting newer patterns and tooling
```

### Rationale behind conservative policy of breaking changes

* Minimize disruption for end users
* Be mindful of the fact that some users are still running obsolete versions of Node in production

> ### Guidelines for dropping support of Node.js versions

Reviewer: I think this is the section I want us to discuss more deeply. Ideally we also put together a calendar view like Node does to see in comparison to Node majors how express majors would line up.

Reviewer 3: Yeah! Having a calendar is one of the best ways to explain the rules for the end users, but currently is very solid 🙂

Reviewer: I think we should make a diagram like node has, maybe even overlay our support *over* the existing node support schedules?

Author: @wesleytodd @UlisesGascon I think @ljharb strongly objected to having a timed deprecation without a practical justification. We need to have this established unambiguously within the strategy. Do we or not do time schedule-based Node version deprecation?

Reviewer: Can we merge without a consensus on this? I am leaning yes? And then we can open targeted PRs to update these and discuss after?

Reviewer 2: If it’s worded as something like “if a release is warranted, this is the schedule when it will happen - but if not, no release will happen” then my concern is gone.

Reviewer 2: Either way, but it seems easy enough to do it before merging ¯\\\_(ツ)_/¯

Reviewer: @kibertoad if you get to it, awesome. If not, lets try and merge this before EOD so we can start those other PRs to clarify and expand.

Author: addressed. unless there are major outstanding objections to some of the wording, I would suggest merging it as-is and iterate in follow-up PRs.

@wesleytodd Can I get push permissions for the repo?

Author: thanks!

> * In order to smoothen the experience of users updating to a new semver major Express version, Express should always support at least one even numbered Node.js version above the current lowest supported one before the drop can be considered. For example, if the lowest supported version is 20, Express support for Node 20 cannot be dropped until Express supports at least Node 22. This ensures that Express updates can be done independently from the Node.js version update.

Reviewer 2: this is great, thanks for capturing it

> * Version support removal must come with clear benefits for one of the three - Express users, Express maintainers or the Node.js core project. Version support is not dropped on a merely time-based basis.

Reviewer 2: ```suggestion
* Version support removal must come with clear benefits for one of the three - Express users, Express maintainers, or the Node.js core project. Version support is not dropped on a merely time-based basis, or otherwise arbitrarily.
```

* Any discontinuation of a Node.js version support is always a semver major release.

### Active branches

> * Several semver major branches can be maintained at the same time, in case there is a significant userbase for older semver major releases.

Reviewer: ```suggestion
* Several semver major branches will be maintained at the same time while we have overlapping supported versions
```

Reviewer 2: this should probably be done as a matter of course anyways, so that if a security issue pops up that necessitates a backport, all the necessary infra is already in place.

> * No features are backported to older semver major branches

Reviewer: Thumbs up on this, but I do think maybe we want to discuss this as a group in more detail.

Reviewer 3: Yep, we should backport only fixes, perf improvements and so on.. but agree that new features are out of scope 💯

Author: let's revise in a separate PR if needed

> * Critical bugfixes may be backported to older semver major branches

Reviewer: ```suggestion
* Critical bugfixes will be backported to older supported semver major branches
```

I think we should commit to this.

Reviewer 2: i agree, but i think there should also be "may be backported to unsupported semver major branches" as well.

> * High-impact security fixes are very likely to be backported to older semver major branches

Reviewer: ```suggestion
* Security fixes will be backported to older supported semver major branches
```

I also think we should commit to doing this for their support lifetime.
